[BLOCK-1952] - Purchase option use cases (#105)
ultraio/docs-blockchain: docs/guides/Uniq Factory Management/exchange-a-uniq-using-smart-contract.md

---
title: 'Exchange a Uniq Using Smart Contract'
deploy: ['staging', 'mainnet']
order: 3
---

# Exchange a Uniq Using Smart Contract

## What does "Exchange a Uniq" mean

> In this guide we will cover the possibility of issuing a new Uniq to the user when he burns or transfers some other Uniq. The purpose of such an exchange may be to migrate user Uniq to a newer factory which may have different set of rules (e.g. Uniqs from new factory can be transferred).

Reviewer: ```suggestion
In this guide, we will cover the possibility of issuing a new Uniq to the user when he burns or transfers some other Uniq. The purpose of such an exchange may be to migrate a user Uniq to a newer factory which may have different set of rules (e.g. Uniqs from new factory can be transferred).
```

> Alternative use case could be if you want to allow users to redeem a Uniq using some "ticket" Uniq which does not have anything useful by itself but can be exchanged for an actual Uniq from a different factory.

Reviewer: ```suggestion
An alternative use case could be if you want to allow users to redeem a Uniq using some "ticket" Uniq which does not have anything useful by itself but can be exchanged for an actual Uniq from a different factory.
```

> Since some of the values inside the factory cannot be changed after creation that can be an alternative solution to effectively provide the option to users to migrate to a new factory with alternative values.

Reviewer: ```suggestion
Some of the factory's values are immutable after they are created, and when a Uniq is exchanged it provides the opportunity for customers to migrate to a new factory with different values.
```

## When do you need a smart contract

You won't need a dedicated smart contract for each case where you need to exchange Uniqs. In some cases the first-hand purchase options may be sufficient enough: [factory purchase options](./factory-purchase-options.md#purchase-option-use-cases).

You should consider using the smart contract approach in the following scenarios:
- I want to exchange Uniq only if some specific conditions are met which are not covered by first-hand purchase feature
    - examples include: requiring preregistration using a dedicated smart contract action; having alternative pricing model where you will check that the payment was done using smart contract; 
- I have some automation requirements which requires minimum user input

>     - since smart contracts are flexible you are able to do more things than besides simply minting a Uniq

Reviewer: ```suggestion
    - since smart contracts are flexible, you are able to do more things than besides simply minting a Uniq
```

- There is an extra interaction with other smart contract that needs to happen
    - NFT contract by default notifies only certain accounts about the action execution and if you need to extend this list you will have to rely on your own smart contract to notify other contracts

## Overview of the smart contract

Smart contract provided on this page does and showcases the following:
- Smart contract is notified when token is burnt or transferred
    - Burn notification happens only if contract is deployed to the same account as token factory manager
    - Transfer notification happens for both sender, receiver of the token and the token factory manager
- Smart contract issues a new Uniq from the pre-configured factory to the original owner of the Uniq

> To make the contract below work based on your needs you will need to do some adjustments:

Reviewer: ```suggestion
To make the contract below work based on your needs, you will need to do some adjustments:
```

- Change name of the factory manager account to the one appropriate to the factory you are trying to mint from
- Change factory id of the factory you are going to mint from
- Add any necessary preliminary checks before issuing a token
    - you may potentially want to check what token was burnt or transferred (id of the token, serial number of factory id)

>     - there could be other requirements that you want to impose like requiring user to register via some other smart contract action

Reviewer: ```suggestion
    - there could be other requirements that you want to impose, like requiring user to register via some other smart contract action
```

> - Depending on the permission structure and where you put the contract you may want to add `smart_contract_name@eosio.code` permission under the `factory_manager@active` so that the contract will be able to issue tokens in the name of the factory manager

Reviewer: ```suggestion
- Depending on the permission structure and where you put the contract, you may want to add `smart_contract_name@eosio.code` permission under the `factory_manager@active` so that the contract will be able to issue tokens in the name of the factory manager
```
